tests/test_dimension.py

```diff
@@ -819,3 +819,79 @@ def test_implicit_dims(self):
             F[i] = i if i < stop_value else stop_value
 
         assert np.all(f.data == F)
+
+    def test_no_fusion_simple(self):
+        """
+        If ConditionalDimensions are present, then Clusters must not be fused so
+        that ultimately Eqs get scheduled to different loop nests.
+        """
+        grid = Grid(shape=(4, 4, 4))
+        time = grid.time_dim
+
+        f = TimeFunction(name='f', grid=grid)
+        g = Function(name='g', grid=grid)
+        h = Function(name='h', grid=grid)
+
+        # No ConditionalDimensions yet. Will be fused and optimized
+        eqns = [Eq(f.forward, f + 1),
+                Eq(h, f + 1),
+                Eq(g, f + 1)]
+
+        op = Operator(eqns)
+
+        exprs = FindNodes(Expression).visit(op._func_table['bf0'].root)
+        assert len(exprs) == 4
+        assert exprs[1].expr.rhs is exprs[0].output
+        assert exprs[2].expr.rhs is exprs[0].output
+        assert exprs[3].expr.rhs is exprs[0].output
+
+        # Now with a ConditionalDimension. No fusion, no optimization
+        ctime = ConditionalDimension(name='ctime', parent=time, condition=time > 4)
+
+        eqns = [Eq(f.forward, f + 1),
+                Eq(h, f + 1),
+                Eq(g, f + 1, implicit_dims=[ctime])]
+
+        op = Operator(eqns)
+        exprs = FindNodes(Expression).visit(op._func_table['bf0'].root)
+        assert len(exprs) == 3
+        assert exprs[1].expr.rhs is exprs[0].output
+        assert exprs[2].expr.rhs is exprs[0].output
+        exprs = FindNodes(Expression).visit(op._func_table['bf1'].root)
+        assert len(exprs) == 1
+
+    def test_no_fusion_convoluted(self):
+        """
+        Conceptually like `test_no_fusion_simple`, but with more expressions
+        and non-trivial data flow.
+        """
+        grid = Grid(shape=(4, 4, 4))
+        time = grid.time_dim
+
+        f = TimeFunction(name='f', grid=grid)
+        g = Function(name='g', grid=grid)
+        h = Function(name='h', grid=grid)
+
+        ctime = ConditionalDimension(name='ctime', parent=time, condition=time > 4)
+
+        eqns = [Eq(f.forward, f + 1),
+                Eq(h, f + 1),
+                Eq(g, f + 1, implicit_dims=[ctime]),
+                Eq(f.forward, f + 1, implicit_dims=[ctime]),
+                Eq(f.forward, f + 1),
+                Eq(g, f + 1)]
```

Comment:
Not sure I follow, there is duplicate equation on purpose?

Reply:
All it matters is that the RHSs are identical

Semantically this test is stupid.

Practically it's useful to see that guarded and un-guarded Clusters are
1. processed separately
2. optimized as expected

Reply:
Oh ok, it's super confusing to read the equations :D

Reply:
Think of them as statements. 

I agree the example is stupid, but at least it's concise